1_fetch/src/get_nwis_data.R

```diff
@@ -1,19 +1,37 @@
+download_nwis_data <- function(site_nums = c("01427207", "01432160", "01436690", "01466500"), Save_as_csv = FALSE){
 
-download_nwis_data <- function(site_nums = c("01427207", "01432160", "01436690", "01466500")){
+  parameterCd <-  '00010'
+  startDate <-"2014-05-01"
+  endDate <- "2015-05-01"
   
-  # create the file names that are needed for download_nwis_site_data
-  # tempdir() creates a temporary directory that is wiped out when you start a new R session; 
-  # replace tempdir() with "1_fetch/out" or another desired folder if you want to retain the download
-  download_files <- file.path(tempdir(), paste0('nwis_', site_nums, '_data.csv'))
   data_out <- data.frame()
-  # loop through files to download 
-  for (download_file in download_files){
-    download_nwis_site_data(download_file, parameterCd = '00010')
-    # read the downloaded data and append it to the existing data.frame
-    these_data <- read_csv(download_file, col_types = 'ccTdcc')
-    data_out <- bind_rows(data_out, these_data)
-  }
+  
+  for(site_num in site_nums){
+
+    data <- readNWISdata(sites=site_num, service="iv", parameterCd = parameterCd, startDate = startDate, endDate = endDate)
+
+    ## COMMENTED THIS OUT as it breaks the code, tbd why    
+    
+    # -- simulating a failure-prone web-service here, do not edit --
+    set.seed(Sys.time())
+    if (sample(c(T,F,F,F), 1)){
+      stop(site_num, ' has failed due to connection timeout. Try tar_make() again')
+    }
+    
+    data_out <- bind_rows(data_out, data)
+
+    }  
+
+    ## Added in case user wants the save the data extracted from NWIS to local dir, Default is False for now
+  if(Save_as_csv == TRUE){
```

Comment:
This is an interesting addition, and something which could be really nice to have. Often we want to be able to write out some portion of our results midway through a pipeline! I like it.
 
However, the way that it is structured is not pipelines-friendly. Here is why: in order to get this optional code to run you would have to change either the `site_data` target or the `download_nwis_data() function itself. This would force `targets` to run every target downstream of this again! A more pipelines-friendly strategy would be to take this code out and put it in a different target which depends on `site_data` and writes the CSV. That way, you can turn the function on and off without forcing everything else to re-run.

Reply:
Now that I removed the for loop in download_nwis_data() and combine the 4 different nwis datasets in a different function - therefore has a separate target - I move this addition there. Let me know if it works.